diff --git a/conf/atcommands.yml b/conf/atcommands.yml
index 232b28fb8..2af65edf6 100644
--- a/conf/atcommands.yml
+++ b/conf/atcommands.yml
@@ -805,6 +805,9 @@ Body:
   - Command: trade
     Help: |
       Params: <char name> - Open a trade window with a another player
+  - Command: killcounter
+    Aliases:
+      - kc
   - Command: recallall
     Help: |
       Warps every character online to you.
diff --git a/src/map/atcommand.cpp b/src/map/atcommand.cpp
index 22ef142a7..f02cbbb5e 100644
--- a/src/map/atcommand.cpp
+++ b/src/map/atcommand.cpp
@@ -3939,6 +3939,101 @@ ACMD_FUNC(idsearch)
 	return 0;
 }
 
+/*==========================================
++ * name/id amount
++ * message = defaut input then sscanf for forward into var
++ * @killcounter id/name pos
++ * @killcounter reset (pos)
++ * @killcounter status
++ *	(sd) int mobcount[MAX_KILLCOUNTER];
++ *	(sd) int mobcountid[MAX_KILLCOUNTER];
++ *------------------------------------------*/
+ACMD_FUNC(killcounter)
+{
+	char cmdsnd[7], monster[NAME_LENGTH];
+	int mob_id, i, pos;
+
+	nullpo_retr(-1, sd);
+
+	memset(cmdsnd, '\0', sizeof(cmdsnd));
+	memset(monster, '\0', sizeof(monster));
+	memset(atcmd_output, '\0', sizeof(atcmd_output));
+
+	if (!message || !*message) {
+		if (sd->state.killcounter) {
+			sd->state.killcounter = 0;
+			clif_displaymessage(fd, "[OFF] - Monster kill will not being count anymore.");
+		} else {
+			sd->state.killcounter = 1;
+			clif_displaymessage(fd, "[ON] - Monster kill is now counted.");
+		}
+		return 0;
+	}
+
+	if (sscanf(message, "%23s %d", monster, &pos) > 1){
+		if(pos < 1 || pos > MAX_KILLCOUNTER){
+			sprintf(atcmd_output, "You can only choose a number between 1 and %d",MAX_KILLCOUNTER);
+			clif_displaymessage(fd, atcmd_output);
+			return -1;
+		}
+	} else {
+		if(strcmp(message,"reset") == 0){
+			for(i=0;i<MAX_KILLCOUNTER;i++){
+				sd->mobcount[i] = 0;
+			}
+			clif_displaymessage(fd, "Reset of all counters done.");
+		} else if(strcmp(message,"status") == 0){
+			clif_displaymessage(fd, sd->state.killcounter?"[ON] - Kill counter":"[OFF] - Kill counter");
+			for(i=0;i<MAX_KILLCOUNTER;i++){
+				if(sd->mobcountid[i] == 0)
+					sprintf(atcmd_output, "[%d] - No monster has been defined on this positions.", i+1);
+				else {
+					std::shared_ptr<s_mob_db> mob = mob_db.find(sd->mobcountid[i]);
+					if (mob == nullptr)
+						continue;
+					sprintf(atcmd_output, "[%d] [%d] %s - %d monster(s) has been killed.", i+1, sd->mobcountid[i], mob->name.c_str(), sd->mobcount[i]);
+				}
+				clif_displaymessage(fd, atcmd_output);
+			}
+		} else {
+			clif_displaymessage(fd, "@killcounter : activate / disactivate the count");
+			clif_displaymessage(fd, "@killcounter reset : reset all counter");
+			clif_displaymessage(fd, "@killcounter reset <position> : reset the counter on the position");
+			clif_displaymessage(fd, "@killcounter <mobID / mobname> <position> : start to count kill on the chosen position");
+			clif_displaymessage(fd, "@killcounter status :  show the number of monsters killed in all your position");
+		}
+		return 0;
+	}
+
+	if(strcmp(monster,"reset") == 0){
+		sd->mobcount[pos-1] = 0;
+		clif_displaymessage(fd, "Reset of selected counter done.");
+		return 0;
+	}
+
+	if (!sd->state.killcounter) {
+		sd->state.killcounter = 1;
+		clif_displaymessage(fd, "[ON] - Monster kill is now counted.");
+	}
+
+	if ((mob_id = mobdb_searchname(monster)) == 0) // check name first (to avoid possible name begining by a number)
+		mob_id = mobdb_checkid(atoi(monster));
+	if (mob_id == 0) {
+		clif_displaymessage(fd, msg_txt(sd,40)); // Invalid monster ID or name.
+		return -1;
+	}
+
+	sd->mobcountid[pos-1] = mob_id;
+	sd->mobcount[pos-1] = 0;
+
+	std::shared_ptr<s_mob_db> mob = mob_db.find(sd->mobcountid[pos-1]);
+
+	sprintf(atcmd_output, "[%d] [%d] %s - %d monsters has been killed.", pos, sd->mobcountid[pos-1], mob->name.c_str(), sd->mobcount[pos-1]);
+	clif_displaymessage(fd, atcmd_output);
+
+	return 0;
+}
+
 /*==========================================
  * Recall All Characters Online To Your Location
  *------------------------------------------*/
@@ -10791,6 +10886,7 @@ void atcommand_basecommands(void) {
 		ACMD_DEF(idsearch),
 		ACMD_DEF(broadcast), // + /b and /nb
 		ACMD_DEF(localbroadcast), // + /lb and /nlb
+		ACMD_DEF(killcounter),
 		ACMD_DEF(recallall),
 		ACMD_DEFR(reload,ATCMD_NOSCRIPT),
 		ACMD_DEF2("reloaditemdb", reload),
diff --git a/src/map/chrif.cpp b/src/map/chrif.cpp
index 35a9cfd21..9d7b4b462 100644
--- a/src/map/chrif.cpp
+++ b/src/map/chrif.cpp
@@ -311,6 +311,12 @@ int chrif_save(struct map_session_data *sd, int flag) {
 
 	chrif_bsdata_save(sd, ((flag&CSAVE_QUITTING) && !(flag&CSAVE_AUTOTRADE)));
 
+	//killcounter
+	for (int i=0;i<MAX_KILLCOUNTER;i++){
+		pc_setreg2(sd,"#mobcountid"+i, sd->mobcountid[i]);
+		pc_setreg2(sd,"#mobcount"+i, sd->mobcount[i]);
+	}
+
 	if (sd->storage.dirty)
 		storage_storagesave(sd);
 	if (flag&CSAVE_INVENTORY)
diff --git a/src/map/mob.cpp b/src/map/mob.cpp
index bd12630d5..dd8dc8aa9 100644
--- a/src/map/mob.cpp
+++ b/src/map/mob.cpp
@@ -2586,6 +2586,15 @@ int mob_dead(struct mob_data *md, struct block_list *src, int type)
 	if( src && src->type == BL_PC ) {
 		sd = (struct map_session_data *)src;
 		mvp_sd = sd;
+		//killcounter
+		if(sd->state.killcounter){
+			for(i=0;i<MAX_KILLCOUNTER;i++){
+				if(sd->mobcountid[i] == md->mob_id){
+					sd->mobcount[i] = sd->mobcount[i]+1;
+					break;
+				}
+			}
+		}
 	}
 
 	if( md->guardian_data && md->guardian_data->number >= 0 && md->guardian_data->number < MAX_GUARDIANS )
diff --git a/src/map/pc.cpp b/src/map/pc.cpp
index 8e2bec585..17eb30feb 100755
--- a/src/map/pc.cpp
+++ b/src/map/pc.cpp
@@ -1919,6 +1919,12 @@ void pc_reg_received(struct map_session_data *sd)
 	}
 	sd->roulette.prizeIdx = -1;
 
+	//killcounter
+	for (int i=0;i<MAX_KILLCOUNTER;i++){
+		sd->mobcountid[i] = static_cast<int>(pc_readreg2(sd, "#mobcountid"+i));
+		sd->mobcount[i] = static_cast<int>(pc_readreg2(sd, "#mobcount"+i));
+	}
+
 	//SG map and mob read [Komurka]
 	for(i=0;i<MAX_PC_FEELHATE;i++) { //for now - someone need to make reading from txt/sql
 		uint16 j;
diff --git a/src/map/pc.hpp b/src/map/pc.hpp
index 50721d387..4abeda355 100644
--- a/src/map/pc.hpp
+++ b/src/map/pc.hpp
@@ -43,6 +43,7 @@ enum sc_type : int16;
 #define MAX_SERVANTBALL 5 /// Max servant weapons
 #define MAX_SERVANT_SIGN 5 /// Max servant signs
 #define MAX_ABYSSBALL 5 /// Max abyss spheres
+#define MAX_KILLCOUNTER 5 /// Max killcounter save
 
 #define LANGTYPE_VAR "#langtype"
 #define CASHPOINT_VAR "#CASHPOINTS"
@@ -343,6 +344,7 @@ struct map_session_data {
 		unsigned int showdelay :1;
 		unsigned int showexp :1;
 		unsigned int showzeny :1;
+		unsigned int killcounter :1; // killcounter [Shakto]
 		unsigned int noask :1; // [LuzZza]
 		unsigned int trading :1; //[Skotlex] is 1 only after a trade has started.
 		unsigned int deal_locked :2; //1: Clicked on OK. 2: Clicked on TRADE
@@ -410,6 +412,10 @@ struct map_session_data {
 	unsigned int permissions;/* group permissions */
 	int count_rewarp; //count how many time we being rewarped
 
+	//killcounter
+	int mobcount[MAX_KILLCOUNTER] = { 0 };
+	int mobcountid[MAX_KILLCOUNTER] = { 0 };
+
 	int langtype;
 	struct mmo_charstatus status;
 
